p2p/discovery/mdns.go

```diff
@@ -58,9 +61,25 @@ func getDialableListenAddrs(ph host.Host) ([]*net.TCPAddr, error) {
 		if err != nil {
 			continue
 		}
-		tcp, ok := na.(*net.TCPAddr)
-		if ok {
-			out = append(out, tcp)
+
+		// if `iface` is specified, use address from `iface`
+		// otherwise get any TCP-capable address.
+		if iface != nil {
+			ips, _ := iface.Addrs()
+			for _, i := range ips {
+				if strings.Contains(na.String(), strings.Split(i.String(), "/")[0]) {
```

Comment:
Actually, `i` is an `IPNet`. 
Here is the doc: 
```go
// String returns the CIDR notation of n like "192.0.2.0/24"
// or "2001:db8::/48" as defined in RFC 4632 and RFC 4291.
// If the mask is not in the canonical form, it returns the
// string which consists of an IP address, followed by a slash
// character and a mask expressed as hexadecimal form with no
// punctuation like "198.51.100.0/c000ff00".
func (n *IPNet) String() string
``` 
I am not sure I understand the goal of checking the Network as the tcp address is appended anyway?
Regarding the network part, when I run on my workstation, `i.Network()` returns `"ip+net"`, while `na.Network` returns `"tcp"`.

Reply:
`iface ` is a `net.Interface`. The Addrs() call on it looks like:

```
func (ifi *Interface) Addrs() ([]Addr, error) {
```

Where `Addr` is

```
type Addr interface {
	Network() string // name of the network (for example, "tcp", "udp")
	String() string  // string form of address (for example, "192.0.2.1:25", "[2001:db8::1]:80")
}
```

Am I missing something here ?

Given this API, I'm saying we should compare the networks because `na` could be a UDP address for example and `i.String()` could return the same host:port but say for a TCP transport in which case, we shouldn't be using it.

Reply:
:thinking: 
`i` is an `net.IPNet` which doesn't seem to respect the Addr interface documentation, while `na` is `net.TCPAddr`.
```go
log.Info("i string: ", i.(net.Addr).String())
log.Info("i network: ", i.(net.Addr).Network())
log.Info("na string: ", na.(net.Addr).String())
log.Info("na network: ", na.(net.Addr).Network())
```
```
2020-11-16T16:25:24.760+0100    INFO    mdns    i string: 192.168.1.4/24
2020-11-16T16:25:24.760+0100    INFO    mdns    i network: ip+net
2020-11-16T16:25:24.760+0100    INFO    mdns    na string: [::1]:43931
2020-11-16T16:25:24.760+0100    INFO    mdns    na network: tcp
```